### PR TITLE
KVM: introspection: clarify the information sent for some vCPU events

### PR DESCRIPTION
**Pull request by adlazar**
hypercall, breakpoint and page fault events

### CODE REVIEW

**Comment by Wenzel** [comment]
Thanks.

Just a quick thing that I noticed while reading the whole document also:

> This event is sent when a hypervisor page fault occurs due to a failed
> permission check in the shadow page tables, the introspection has been
> permission check in the shadow page tables

Should we use the term `Shadow page tables` ?
To me, this refer to the 2nd level of page tables emulated by the hypervisor (without hardware assistance like EPT/NPT)
As KVMi is only implemented when EPT/NPT are available, should we change the vocabulary here maybe ?


I'm not 100% sure here though.
What do you think ?

**Comment by adlazar** [comment]
I've removed these words with the second commit.

File: Documentation/virt/kvm/kvmi.rst [hunk 1]
```diff
@@ -1819,20 +1829,33 @@ sent to the introspection tool. The *CONTINUE* action will set the ``new_val``.
...
+(eg. KVMI_PAGE_ACCESS_R) are sent to the introspection tool.
+
+In case of a restricted read access, the guest address is the location
+of the memory being read. On write access, the guest address is the
+location of the memory being written. On execute access, the guest
+address is the location of the instruction being executed.
+
+In the current implementation, most of these events are sent during
+emulation. If the page fault has set more than on access bit (e.g. r-x/-rw),
+the introspection tool may receive more then one KVMI_PF_EVENT
```
Comment (Wenzel) [inline Documentation/virt/kvm/kvmi.rst]: more `than` one ?

File: Documentation/virt/kvm/kvmi.rst [hunk 2]
```diff
@@ -1819,20 +1829,33 @@ sent to the introspection tool. The *CONTINUE* action will set the ``new_val``.
...
 introspection tool the chance to audit the code attempting the operation.
 
-``kvmi_event``, guest virtual address (or 0xffffffff/UNMAPPED_GVA),
-guest physical address and the access flags (eg. KVMI_PAGE_ACCESS_R)
-are sent to the introspection tool.
+``kvmi_event`` (with the vCPU state), guest virtual address (or
+-1/UNMAPPED_GVA), guest physical address and the access flags
+(eg. KVMI_PAGE_ACCESS_R) are sent to the introspection tool.
+
+In case of a restricted read access, the guest address is the location
```
Comment (Wenzel) [inline Documentation/virt/kvm/kvmi.rst:1849]: Maybe being specific about the struct field we are talking about would help ?

the guest address (`kvmi_event_pf.gva`) ...

You could also add info about `regs.rip`:
In this context, the `kvmi_event->event.common.arch.regs.rip` contains the guest vrtual address of the instruction responsible for the memory read / write

What do you think ?

**Review by Wenzel — COMMENTED** [review]
I left a few comments, what do you think.

File: Documentation/virt/kvm/kvmi.rst (hunk 2 above)
Comment (adlazar) [reply Documentation/virt/kvm/kvmi.rst:1849]: Sure.

**Pull request merged**